### PR TITLE
Feature/szewczyk/tlv api enhancements

### PR DESCRIPTION
**Pull request by robszewczyk**
Three minor TLV API enhancements:

* A new constructor to initialize `WeaveCircularTLVBuffer` to start at any offset within the buffer
* A new flavor of `nl::Weave::TLV::Utilities::Find` that is able to find elements based on a predicate
* Added a missing accessor for the underlying buffer in the `TLVReader`

These changes were motivated by the restructuring of the external event storage within event logging

### CODE REVIEW

**Comment by gerickson** [comment]
> > Code looks good; what about unit tests?
> 
> Added unit tests, made the API parallel with other parts of the TLVUtilities, fixed up doxygen

Thanks! Code looks good; hopefully, CI failures are trivial.

File: src/lib/core/WeaveTLV.h [hunk 1]
```diff
@@ -141,6 +141,7 @@ friend class TLVUpdater;
     uint32_t GetRemainingLength(void) const { return mMaxLen - mLenRead; }
 
     const uint8_t *GetReadPoint(void) const { return mReadPoint; }
+    uintptr_t GetBufHandle(void) const { return mBufHandle; }
```
Comment (jaylogue) [inline src/lib/core/WeaveTLV.h]: Out of curiosity, is there a use case for this?

File: src/lib/core/WeaveCircularTLVBuffer.cpp [hunk 2]
```diff
@@ -49,7 +49,33 @@ using namespace nl::Weave::Encoding;
  *
  * @param[in] inBuffer       A pointer to the backing store for the queue
  *
- * @param[in] inBufferLength Length of the backing store
+ * @param[in] inBufferLength Length, in bytes, of the backing store
+ *
+ * @param[in] inHead         Initial point for the head.  The @a inHead pointer is must fall within the backting store for the circular buffer, i.e. within @a inBuffer and &(@a inBuffer[@a inBufferLength])
```
Comment (Glen-Weaver) [inline src/lib/core/WeaveCircularTLVBuffer.cpp]: "backting" -> "backing"

File: src/lib/core/WeaveTLVUtilities.cpp [hunk 3]
```diff
@@ -336,6 +336,130 @@ WEAVE_ERROR Find(const TLVReader &aReader, const uint64_t &aTag, TLVReader &aRes
 
     return retval;
 }
+/**
+ *  Search for the specified tag within the provided TLV reader,
```
Comment (Glen-Weaver) [inline src/lib/core/WeaveTLVUtilities.cpp]: This comment block does not represent the following code.

File: src/lib/core/WeaveTLV.h (hunk 1 above)
Comment (robszewczyk) [reply src/lib/core/WeaveTLV.h]: It's something I'm not particularly happy with: I needed to create a brand new circular buffer and associated reader within the storage used by an existing reader.

File: src/lib/core/WeaveTLVUtilities.cpp (hunk 3 above)
Comment (robszewczyk) [reply src/lib/core/WeaveTLVUtilities.cpp]: Thanks, removed.

File: src/lib/core/WeaveCircularTLVBuffer.cpp (hunk 2 above)
Comment (robszewczyk) [reply src/lib/core/WeaveCircularTLVBuffer.cpp]: Thanks, edited.

**Pull request merged**